staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml

```diff
@@ -0,0 +1,18 @@
+[[constraint]]
+  branch = "master"
+  name = "k8s.io/apimachinery"
+
+[[constraint]]
+  branch = "master"
+  name = "k8s.io/apiserver"
+
+[[constraint]]
+  name = "k8s.io/client-go"
+  version = "6.0.0"
```

Comment:
where does this file come from? Is it manually written?

Asking because this must be kept up-to-date. At least we need some hack/verify-* script to check that no constraint is missing

Do we really need the branch and version constraints here? Won't your dep fork figure out that the staging packages are in the same repo?

Background: on publishing we will drop this file (via https://github.com/kubernetes/publishing-bot/blob/master/configs/kubernetes-rules-configmap.yaml#L9) for now. 

I tried to keep the file in https://github.com/kubernetes/publishing-bot/pull/55 with full dependencies and hit a wall with the solver. Maybe we can do a minimal file like you have here. But it has to be generated by the bot automatically from the branch config.

Reply:
Sorry, those are leftovers from previous iterations. I think at this point, those should mostly be empty and it's probably easier for all constraints to be expressed in the main Gopkg.toml anyway.
I guess I could implement some smarts that would replicate the constraints that make sense in each sub-project, but I'm not convinced that's needed for now.

I suppose I should consider generating the corresponding Gopkg.lock files though...

Reply:
btw my experience is that it's easier to generate a Gopkg.toml using overrides rather than constraints when you know ahead of time what the desired solution is.
Of course, the result is not flexible, but since it'll be re-generated when needed anyway, it's probably no big deal.

Reply:
> rather than constraints when you know ahead of time what the desired solution is.

And overrides are ignored when they are in a library. That's the reason they are of no use in client-go. Also the lock files is ignored.

Reply:
To help users with dep though, we could create a Gopkg.toml.template file. They can copy that into their projects, with a long list of overrides. Wdyt? Better than nothing.

Reply:
Oh sorry I misunderstood the problem at hand.
Yeah I guess providing a template might help somewhat, although it just pushes the problem one step further, as you note.

Reply:
btw I should add that it's entirely possible that the current "solution" expressed in Godeps.json is not an actual solution and that packages would conflict with each other according to their respective constraints (which overrides casually ignore).
I don't know if that's the case currently, but a few weeks back, there was a clear inconsistency for some microsoft-related packages (whether the corresponding constraints were faulty or too broad is another question :))

My point is, in general we shouldn't assume that a godep-snapshot is dep-consistent, so generating one from the other is a risky endeavor by nature.

Reply:
Godep does no consistency checks. It only checks out what you declare in Godeps.json. So if kubernetes itself (all of it) builds and works with the given revisions, so will the staging repo.

If I understand you correctly, you are referring to constraints in dependencies' Gopkg.toml. The Godep.json solution might now matchs of the dependencies in this sense. This of course can happen.

Do I understand it correctly that the generated Godeps.jsons of all staging repos agree on the revisions? I.e. you don't try to find solutions for each staging dir alone, but one global one.

Reply:
Right, yes all generated here `Godeps.json` are consistent, and use the versions from the unified `Gopkg.lock`

What I'm trying to express is that this `Gopkg.lock` solution is not necessarily a "real" one, in the sense that it's not the result of the resolution of proper constraints (from kubernetes itself or any of the vendored repository that happens to have a `Gopkg.toml` too). Instead, it's a forced replica (via overrides) of whatever the current `godep` solution is, at the time of the conversion.

As such, there can be no expectation (yet) that using any of the kubernetes repositories in the context of a proper dep-managed downstream project (e.g. vendoring `client-go`) should lead to dep finding a solution of said downstream project since there might not even be a dep-computable solution for the repository itself at this point. And that's also what could happen with your solver problem above  (kubernetes/publishing-bot#55)

That's a problem that should gradually disappear after migration to dep, when we replace overrides with constraints, make the dependencies versions evolve, and hopefully converge towards an override-free manifest.
In short, as long as overrides are needed, downstream is screwed and needs to override too :( (just as it is today)